cmd/docker/docker.go

```diff
@@ -135,32 +142,36 @@ func dockerPreRun(opts *cliflags.ClientOptions) {
 	}
 }
 
-func hideUnsupportedFeatures(cmd *cobra.Command, clientVersion string, hasExperimental bool) {
+func hideUnsupportedFeatures(cmd *cobra.Command, apiProps apiProperties) {
 	cmd.Flags().VisitAll(func(f *pflag.Flag) {
 		// hide experimental flags
-		if !hasExperimental {
+		if !apiProps.Experimental {
 			if _, ok := f.Annotations["experimental"]; ok {
 				f.Hidden = true
 			}
 		}
 
 		// hide flags not supported by the server
-		if !isFlagSupported(f, clientVersion) {
+		if !isFlagSupported(f, apiProps.ClientVer) {
+			f.Hidden = true
+		}
+
+		if flagOS, ok := f.Annotations["os"]; ok && len(flagOS) == 1 && flagOS[0] != apiProps.OSType {
```

Comment:
This logic should take older daemons into account; older daemons won't return an `OS-Type` header. As a result, this logic will hide _any_ flag that has a `os` annotation set. If the daemon didn't return an `OS-Type` header, just show the flag.